controllers/gce/firewalls/firewalls.go

```diff
@@ -141,3 +142,53 @@ func (fr *FirewallRules) createFirewallObject(firewallName, description string,
 		TargetTags: targetTags,
 	}, nil
 }
+
+func (fr *FirewallRules) createFirewall(f *compute.Firewall) error {
+	err := fr.cloud.CreateFirewall(f)
+	if utils.IsForbiddenError(err) && fr.cloud.OnXPN() {
+		gcloudCmd := gce.FirewallToGCloudCreateCmd(f, fr.cloud.NetworkProjectID())
+		glog.V(3).Infof("Could not create L7 firewall on XPN cluster. Raising event for cmd: %q", gcloudCmd)
+		return newFirewallXPNError(err, gcloudCmd)
+	}
+	return err
+}
+
+func (fr *FirewallRules) updateFirewall(f *compute.Firewall) error {
+	err := fr.cloud.UpdateFirewall(f)
+	if utils.IsForbiddenError(err) && fr.cloud.OnXPN() {
+		gcloudCmd := gce.FirewallToGCloudUpdateCmd(f, fr.cloud.NetworkProjectID())
+		glog.V(3).Infof("Could not update L7 firewall on XPN cluster. Raising event for cmd: %q", gcloudCmd)
+		return newFirewallXPNError(err, gcloudCmd)
+	}
+	return err
+}
+
+func (fr *FirewallRules) deleteFirewall(name string) error {
+	err := fr.cloud.DeleteFirewall(name)
+	if utils.IsNotFoundError(err) {
+		glog.Infof("Firewall with name %v didn't exist when attempting delete.", name)
+		return nil
+	} else if utils.IsForbiddenError(err) && fr.cloud.OnXPN() {
+		gcloudCmd := gce.FirewallToGCloudDeleteCmd(name, fr.cloud.NetworkProjectID())
+		glog.V(3).Infof("Could not attempt delete of L7 firewall on XPN cluster. %q needs to be ran.", gcloudCmd)
+		// FirewallXPN error is not returned here because there is no ingress for attaching.
```

Comment:
It should return an error irrespective? The caller knows that there is no ingress and hence it will just log it